src/components/CampaignsAdmin/edit/CampaignDetails.tsx

```diff
@@ -1,26 +1,139 @@
+import { useState, useMemo } from 'react';
 import { v4 as uuidv4 } from 'uuid';
-import { DeleteOutlined } from '@ant-design/icons';
-import { Table, Divider, Image, message, Popconfirm, Button } from 'antd';
+import { DeleteOutlined, SearchOutlined, ClearOutlined, PlusOutlined } from '@ant-design/icons';
+import { Table, Image, message, Popconfirm, Button, Modal, Input, PageHeader } from 'antd';
 import { useParams } from 'react-router-dom';
 import { HTML5Backend } from 'react-dnd-html5-backend';
 import { DndProvider } from 'react-dnd';
 
 import { useCampaignById } from 'db/hooks/getCampaignDetailsByID';
+import { useLocations } from 'db/hooks/getLocations';
 
 import { CampaignLocationInfo } from 'models/CampaignLocationInfo';
 import { getSupabaseClient } from 'db/DatabaseClient';
 import { DueDateLabel } from 'components/CampaignsAdmin/DueDateLabel';
 import { DropPhotoZone } from 'components/CampaignsAdmin/DropPhotoZone';
 
+import { ColumnsType } from 'antd/lib/table';
+import { FilterDropdownProps } from 'antd/lib/table/interface';
+
 import 'styles/pages/campaignDetails.scss';
+import { definitions } from 'db/SupabaseTypes';
+
 
 type CampaignParams = {
   id: string;
 };
 
 export function CampaignDetails() {
   const { id } = useParams<CampaignParams>();
-  const { campaign } = useCampaignById(Number(id));
+  const { campaign, refreshCampaign } = useCampaignById(Number(id));
+  const { locations } = useLocations();
+  const [isModalVisible, setIsModalVisible] = useState(false);
+
+  const handleAddLocation = async (newLocation: definitions['Locations']) => {
+    campaign?.locationInfo.push({ ...newLocation, campaignPhotoUrl: '' });
+    const update = await updateLocationConfig(Number(id), campaign?.locationInfo);
+  };
+
+  const showAddLocationModal = () => {
+    setIsModalVisible(true);
+  };
+  const handleModalOk = () => {
+    setIsModalVisible(false);
+  };
+
+  const handleModalCancel = () => {
+    setIsModalVisible(false);
+  };
+  const actions = useMemo(() => {
+    const arr = [
+      <Button type="primary" onClick={showAddLocationModal} key='ARCHIVE'>ADD LOCATION</Button>,
+    ];
+    return arr;
+  }, [campaign]);
+  const addLocationColumns: ColumnsType<definitions['Locations']> = [
```

Comment:
This should be memoized as well. But perhaps is too much code within the same controller and we can move it outside. Let's look at this together whenever you're back